# <center>Sales Analysis </center>

The following analysis responds to a Fictitious sales data for a certain company across the US regions. The project involves conducting a comprehensive analysis of sales data. The analysis aims to answer various questions across different aspects of the business, including sales, customers, products, warehouses, time-series trends, and stores. 

Tools: 
1. SQL
2. Python

In [58]:
#Import the dependencies
import pandas as pd
import matplotlib.pyplot as plt
import sqlalchemy

In [59]:
#Load data
df = pd.read_excel('../US_Regional_Sales_Data.xlsx') 

In [60]:
#create a connection with the PostgreSQL database
from sqlalchemy import create_engine

# create a connection to the database
engine = create_engine('postgresql://postgres:5403@localhost:5432/salesdb')  # replace with your database details

# write the DataFrame to a table in the SQL database
df.to_sql('sales', engine, if_exists='replace')  # replace 'table_name' with your desired table name

991

In [61]:
# Load the SQL extension in Jupyter Notebook
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [62]:
#Connection
%sql postgresql://postgres:5403@localhost:5432/salesdb

In [63]:
# Check if it is working well
%sql SELECT * FROM sales LIMIT 5;


 * postgresql://postgres:***@localhost:5432/salesdb
5 rows affected.


[(0, 'SO - 000101', 'In-Store', 'WARE-UHY1004', datetime.datetime(2017, 12, 31, 0, 0), datetime.datetime(2018, 5, 31, 0, 0), datetime.datetime(2018, 6, 14, 0, 0), datetime.datetime(2018, 6, 19, 0, 0), 'USD', 6, 15, 259, 12, 5, 0.075, 1963.1000000000001, 1001.181),
 (1, 'SO - 000102', 'Online', 'WARE-NMK1003', datetime.datetime(2017, 12, 31, 0, 0), datetime.datetime(2018, 5, 31, 0, 0), datetime.datetime(2018, 6, 22, 0, 0), datetime.datetime(2018, 7, 2, 0, 0), 'USD', 14, 20, 196, 27, 3, 0.075, 3939.6, 3348.66),
 (2, 'SO - 000103', 'Distributor', 'WARE-UHY1004', datetime.datetime(2017, 12, 31, 0, 0), datetime.datetime(2018, 5, 31, 0, 0), datetime.datetime(2018, 6, 21, 0, 0), datetime.datetime(2018, 7, 1, 0, 0), 'USD', 21, 16, 213, 16, 1, 0.05, 1775.5, 781.22),
 (3, 'SO - 000104', 'Wholesale', 'WARE-NMK1003', datetime.datetime(2017, 12, 31, 0, 0), datetime.datetime(2018, 5, 31, 0, 0), datetime.datetime(2018, 6, 2, 0, 0), datetime.datetime(2018, 6, 7, 0, 0), 'USD', 28, 48, 107, 23, 8, 0.075, 2324.9, 1464.6870000000001),
 (4, 'SO - 000105', 'Distributor', 'WARE-NMK1003', datetime.datetime(2018, 4, 10, 0, 0), datetime.datetime(2018, 5, 31, 0, 0), datetime.datetime(2018, 6, 16, 0, 0), datetime.datetime(2018, 6, 26, 0, 0), 'USD', 22, 49, 111, 26, 8, 0.1, 1822.4, 1476.1440000000002)]

### <center>Sales Analysis </center>

#### What is the total sales volume and sales value by Sales Channel?

In [64]:
%%sql

SELECT 
    "Sales Channel",
    SUM("Order Quantity") AS Total_Sales_Volume,
    SUM("Order Quantity" * "Unit Price") AS Total_Sales_Value
FROM 
    sales
GROUP BY 
    "Sales Channel";


 * postgresql://postgres:***@localhost:5432/salesdb
4 rows affected.


[('Online', Decimal('10897'), 24629756.099999994),
 ('Distributor', Decimal('6287'), 14809907.799999993),
 ('Wholesale', Decimal('4100'), 9212948.900000002),
 ('In-Store', Decimal('14878'), 34040113.79999999)]

Insights:

1. The "Online" sales channel has the highest total sales volume with 10,897 units sold. It also has the highest total sales value of 24,629,756.10, indicating a strong performance in terms of both volume and value.
2. The "In-Store" sales channel has the highest total sales volume among all channels with 14,878 units sold. It also has a significant total sales value of 34,040,113.80, making it a key contributor to overall sales.
3. The "Distributor" sales channel has a total sales volume of 6,287 units and a total sales value of 14,809,907.80. Although it has a lower sales volume compared to the "Online" and "In-Store" channels, it still contributes significantly to the overall sales value.
4. The "Wholesale" sales channel has the lowest total sales volume with 4,100 units sold. However, it still generates a substantial total sales value of 9,212,948.90.


#### Which Sales Channel provides the highest average Unit Price and Unit Cost?

In [65]:
%%sql

SELECT 
    "Sales Channel",
    AVG("Unit Price") AS Average_Unit_Price,
    AVG("Unit Cost") AS Average_Unit_Cost
FROM 
    sales
GROUP BY 
    "Sales Channel"
ORDER BY 
    Average_Unit_Price DESC,
    Average_Unit_Cost DESC
LIMIT 1;


 * postgresql://postgres:***@localhost:5432/salesdb
1 rows affected.


[('Distributor', 2328.554545454545, 1456.116603636365)]

The average unit price for the "Distributor" sales channel is approximately $2,328.55. This indicates that, on average, products sold through this channel have a higher price compared to other sales channels. It suggests that the distributor sales channel may be targeting higher-end products or catering to a customer segment willing to pay a premium price.

The average unit cost for the "Distributor" sales channel is approximately $1,456.12. This represents the average cost incurred by the business to procure or produce the products sold through this channel. Comparing it with the average unit price, we can observe that there is a significant margin between the cost and price per unit. This suggests that the distributor sales channel is potentially generating good profitability for the business.

#### Which Sales TeamID generates the most revenue and highest number of orders?

In [66]:
%%sql

SELECT 
    "_SalesTeamID",
    COUNT("OrderNumber") AS Total_Number_of_Orders,
    SUM("Order Quantity" * "Unit Price") AS Total_Revenue
FROM 
    sales
GROUP BY 
    "_SalesTeamID"
ORDER BY 
    Total_Revenue DESC,
    Total_Number_of_Orders DESC
LIMIT 1;



 * postgresql://postgres:***@localhost:5432/salesdb
1 rows affected.


[(26, 296, 3346569.599999999)]

 Sales Team 26 is a significant contributor to the overall revenue and order count. The team's high performance in terms of both revenue and number of orders highlights their effectiveness in generating sales and driving business growth

#### How much discount has been provided by each Sales TeamID?

In [67]:
%%sql

SELECT 
    "_SalesTeamID",
     SUM("Order Quantity" * "Unit Price" * "Discount Applied") AS Total_Discount
FROM 
    sales
GROUP BY 
    "_SalesTeamID"
ORDER BY 
    Total_Discount DESC;

 * postgresql://postgres:***@localhost:5432/salesdb
28 rows affected.


[(1, 435669.6774999998),
 (13, 423124.0950000001),
 (12, 394739.37749999994),
 (19, 374419.95249999966),
 (8, 371546.82500000007),
 (11, 370328.42999999993),
 (2, 368569.3450000003),
 (26, 366156.17249999987),
 (18, 357588.37999999983),
 (9, 355868.99250000017),
 (17, 349712.52999999997),
 (15, 348162.6524999999),
 (21, 342824.5949999998),
 (16, 342421.2550000001),
 (24, 342149.5699999998),
 (6, 341367.8474999999),
 (22, 329730.95249999996),
 (14, 323045.525),
 (20, 319373.7574999999),
 (25, 318494.5500000001),
 (5, 318297.9049999998),
 (7, 315448.22750000004),
 (23, 314276.5649999999),
 (28, 308715.565),
 (10, 294514.245),
 (3, 291735.25249999994),
 (4, 266744.58750000014),
 (27, 264320.025)]

1. The discount provided by each Sales TeamID varies significantly.
2. Sales Team 1 has provided the highest total discount, amounting to \$435,669.68.
3. Sales Team 13 closely follows with a total discount of \$423,124.10.
4. The top five Sales Teams with the highest total discounts are Sales Team 1, 13, 12, 19, and 8.

#### What is the average order quantity per Sales Channel and Sales TeamID?

In [68]:
%%sql

SELECT
    "Sales Channel"
    "_SalesTeamID",
    AVG("Order Quantity") AS Average_Order_Quantity
FROM 
    sales
GROUP BY 
    "Sales Channel",
    "_SalesTeamID"
ORDER BY 
    Average_Order_Quantity DESC;

 * postgresql://postgres:***@localhost:5432/salesdb
31 rows affected.


[('Wholesale', Decimal('4.8097165991902834')),
 ('In-Store', Decimal('4.7967479674796748')),
 ('Wholesale', Decimal('4.6993243243243243')),
 ('Distributor', Decimal('4.6956521739130435')),
 ('Distributor', Decimal('4.6891891891891892')),
 ('In-Store', Decimal('4.6884615384615385')),
 ('Distributor', Decimal('4.6830601092896175')),
 ('Online', Decimal('4.6708860759493671')),
 ('Online', Decimal('4.6321839080459770')),
 ('In-Store', Decimal('4.6317567567567568')),
 ('Distributor', Decimal('4.6278195488721805')),
 ('In-Store', Decimal('4.5605536332179931')),
 ('In-Store', Decimal('4.5333333333333333')),
 ('Online', Decimal('4.5313653136531365')),
 ('Online', Decimal('4.5236363636363636')),
 ('In-Store', Decimal('4.5231788079470199')),
 ('Distributor', Decimal('4.4873646209386282')),
 ('Online', Decimal('4.4865771812080537')),
 ('In-Store', Decimal('4.4730158730158730')),
 ('In-Store', Decimal('4.4567901234567901')),
 ('Online', Decimal('4.4163822525597270')),
 ('Online', Decimal('4.4144736842105263')),
 ('In-Store', Decimal('4.4134275618374558')),
 ('Wholesale', Decimal('4.4078947368421053')),
 ('Online', Decimal('4.4018264840182648')),
 ('In-Store', Decimal('4.3938356164383562')),
 ('In-Store', Decimal('4.3886792452830189')),
 ('Distributor', Decimal('4.3802816901408451')),
 ('Online', Decimal('4.3352941176470588')),
 ('Wholesale', Decimal('4.3284671532846715')),
 ('In-Store', Decimal('4.3102310231023102'))]

#### What is the trend of total sales over time (year, month)?


In [69]:
%%sql

SELECT 
    EXTRACT(YEAR FROM "OrderDate") AS Order_Year,
    EXTRACT(MONTH FROM "OrderDate") AS Order_Month,
    SUM("Order Quantity" * "Unit Price") AS Total_Sales
FROM 
    sales
GROUP BY 
    Order_Year, Order_Month
ORDER BY 
    Order_Year, Order_Month;


 * postgresql://postgres:***@localhost:5432/salesdb
32 rows affected.


[(Decimal('2018'), Decimal('5'), 75629.59999999999),
 (Decimal('2018'), Decimal('6'), 2454752.7),
 (Decimal('2018'), Decimal('7'), 2707550.4),
 (Decimal('2018'), Decimal('8'), 2909421.399999998),
 (Decimal('2018'), Decimal('9'), 2798194.7000000016),
 (Decimal('2018'), Decimal('10'), 2323720.8000000003),
 (Decimal('2018'), Decimal('11'), 2977607.299999998),
 (Decimal('2018'), Decimal('12'), 3046617.2999999984),
 (Decimal('2019'), Decimal('1'), 3075112.3999999985),
 (Decimal('2019'), Decimal('2'), 2168568.9000000004),
 (Decimal('2019'), Decimal('3'), 2199509.5),
 (Decimal('2019'), Decimal('4'), 2844987.4999999995),
 (Decimal('2019'), Decimal('5'), 2795126.099999999),
 (Decimal('2019'), Decimal('6'), 2596986.9999999995),
 (Decimal('2019'), Decimal('7'), 2295285.9999999995),
 (Decimal('2019'), Decimal('8'), 2652918.6000000006),
 (Decimal('2019'), Decimal('9'), 2610427.199999999),
 (Decimal('2019'), Decimal('10'), 2496795.199999999),
 (Decimal('2019'), Decimal('11'), 3065719.000000001),
 (Decimal('2019'), Decimal('12'), 2733352.1),
 (Decimal('2020'), Decimal('1'), 3129998.8000000017),
 (Decimal('2020'), Decimal('2'), 2762182.200000001),
 (Decimal('2020'), Decimal('3'), 2314119.6999999997),
 (Decimal('2020'), Decimal('4'), 2489539.1000000015),
 (Decimal('2020'), Decimal('5'), 2781726.100000001),
 (Decimal('2020'), Decimal('6'), 2349086.9999999986),
 (Decimal('2020'), Decimal('7'), 3108042.899999999),
 (Decimal('2020'), Decimal('8'), 2486758.6),
 (Decimal('2020'), Decimal('9'), 2380630.6),
 (Decimal('2020'), Decimal('10'), 2778717.8000000003),
 (Decimal('2020'), Decimal('11'), 2629475.3),
 (Decimal('2020'), Decimal('12'), 2654164.8000000003)]

1. The average order quantity varies across different Sales Channels and Sales TeamIDs.
2. The "Wholesale" Sales Channel has the highest average order quantity, with values ranging from approximately 4.328 to 4.810. This suggests that customers who purchase through the Wholesale channel tend to place orders with a higher quantity of products on average.
3. The "In-Store" Sales Channel also exhibits relatively high average order quantities, ranging from approximately 4.310 to 4.797. This indicates that customers who make purchases in-store tend to buy a higher quantity of products per order, on average.
4. Other Sales Channels, such as "Distributor" and "Online," also show average order quantities ranging from approximately 4.335 to 4.689.

#### During which months do the highest sales occur?


In [70]:
%%sql

SELECT 
    EXTRACT(MONTH FROM "OrderDate") AS Order_Month,
    AVG("Order Quantity" * "Unit Price") AS Average_Sales
FROM 
    sales
GROUP BY 
    Order_Month
ORDER BY 
    average_sales DESC;


 * postgresql://postgres:***@localhost:5432/salesdb
12 rows affected.


[(Decimal('11'), 11104.739564660686),
 (Decimal('1'), 10905.292091388397),
 (Decimal('12'), 10662.62225031605),
 (Decimal('9'), 10455.372483221478),
 (Decimal('5'), 10277.239636363629),
 (Decimal('7'), 10202.363899371074),
 (Decimal('10'), 10200.313825503354),
 (Decimal('8'), 10175.851580278126),
 (Decimal('3'), 10052.626280623606),
 (Decimal('4'), 10008.492682926826),
 (Decimal('6'), 10001.117162162167),
 (Decimal('2'), 9822.21334661355)]

1. The highest sales occur in the months of November (order_month 11), January (order_month 1), and December (order_month 12), based on the average sales values.
2. November has the highest average sales, followed closely by January and December.
3. The months with the highest average sales are typically associated with increased sales volume and revenue.

#### How many unique stores (_StoreID) do we have?

In [71]:
%%sql

SELECT 
    COUNT(DISTINCT "_StoreID") AS Unique_Stores
FROM 
    sales;


 * postgresql://postgres:***@localhost:5432/salesdb
1 rows affected.


[(367,)]

#### Which store generates the most sales in terms of volume and value?

In [72]:
%%sql

SELECT 
    "_StoreID",
    SUM("Order Quantity") AS Total_Sales_Volume,
    SUM("Order Quantity" * "Unit Price") AS Total_Sales_Value
FROM 
    sales
GROUP BY 
    "_StoreID"
ORDER BY 
    Total_Sales_Value DESC,
    Total_Sales_Volume DESC
LIMIT 1;


 * postgresql://postgres:***@localhost:5432/salesdb
1 rows affected.


[(284, Decimal('182'), 490654.39999999997)]

####  How do sales and order quantities compare among the top ten stores? 

In [73]:
%%sql

SELECT 
    "_StoreID",
    SUM("Order Quantity") AS Total_Order_Quantity,
    SUM("Order Quantity" * "Unit Price") AS Total_Sales
FROM 
    sales
GROUP BY 
    "_StoreID"
ORDER BY 
    Total_Sales DESC,
    Total_Order_Quantity DESC
LIMIT 10;


 * postgresql://postgres:***@localhost:5432/salesdb
10 rows affected.


[(284, Decimal('182'), 490654.39999999997),
 (26, Decimal('178'), 443841.5),
 (238, Decimal('128'), 424713.00000000006),
 (328, Decimal('169'), 393129.19999999995),
 (166, Decimal('174'), 389658.60000000003),
 (33, Decimal('153'), 389605.0),
 (347, Decimal('136'), 387802.6999999999),
 (303, Decimal('145'), 387735.7),
 (100, Decimal('108'), 380117.8),
 (138, Decimal('111'), 366952.3)]

#### Which Warehouse has the highest number of orders?


In [74]:
%%sql

SELECT 
    "WarehouseCode",
    COUNT("OrderNumber") AS Total_Orders
FROM 
    sales
GROUP BY 
    "WarehouseCode"
ORDER BY 
    Total_Orders DESC
LIMIT 1;


 * postgresql://postgres:***@localhost:5432/salesdb
1 rows affected.


[('WARE-NMK1003', 2505)]

#### Do Warehouses affect the delivery time (difference between OrderDate and DeliveryDate)?

In [75]:
%%sql

SELECT 
    "WarehouseCode",
    AVG("DeliveryDate" - "OrderDate") AS Average_Delivery_Time
FROM 
    sales
GROUP BY 
    "WarehouseCode"
ORDER BY 
    Average_Delivery_Time DESC;


 * postgresql://postgres:***@localhost:5432/salesdb
6 rows affected.


[('WARE-NMK1003', datetime.timedelta(days=20, seconds=83226, microseconds=826347)),
 ('WARE-UHY1004', datetime.timedelta(days=20, seconds=65978, microseconds=181818)),
 ('WARE-NBV1002', datetime.timedelta(days=20, seconds=55891, microseconds=172214)),
 ('WARE-PUJ1005', datetime.timedelta(days=20, seconds=47993, microseconds=383873)),
 ('WARE-XYS1001', datetime.timedelta(days=20, seconds=34220, microseconds=621931)),
 ('WARE-MKL1006', datetime.timedelta(days=20, seconds=26313, microseconds=185531))]

These results indicate that the different warehouses involved in the sales process do not have a significant impact on the overall delivery time. It suggests that the delivery process is consistent and efficient across all warehouses, resulting in a similar average delivery time of approximately 20 days. 

#### How do warehouses correlate with specific Sales Channels?

In [76]:
%%sql

SELECT 
    "WarehouseCode", 
    "Sales Channel", 
    COUNT("OrderNumber") AS Total_Orders
FROM 
    sales
GROUP BY 
    "WarehouseCode", "Sales Channel"
ORDER BY 
    "WarehouseCode", Total_Orders DESC;


 * postgresql://postgres:***@localhost:5432/salesdb
24 rows affected.


[('WARE-MKL1006', 'In-Store', 367),
 ('WARE-MKL1006', 'Online', 250),
 ('WARE-MKL1006', 'Distributor', 155),
 ('WARE-MKL1006', 'Wholesale', 85),
 ('WARE-NBV1002', 'In-Store', 291),
 ('WARE-NBV1002', 'Online', 215),
 ('WARE-NBV1002', 'Distributor', 112),
 ('WARE-NBV1002', 'Wholesale', 73),
 ('WARE-NMK1003', 'In-Store', 1029),
 ('WARE-NMK1003', 'Online', 753),
 ('WARE-NMK1003', 'Distributor', 428),
 ('WARE-NMK1003', 'Wholesale', 295),
 ('WARE-PUJ1005', 'In-Store', 599),
 ('WARE-PUJ1005', 'Online', 426),
 ('WARE-PUJ1005', 'Distributor', 262),
 ('WARE-PUJ1005', 'Wholesale', 164),
 ('WARE-UHY1004', 'In-Store', 517),
 ('WARE-UHY1004', 'Online', 388),
 ('WARE-UHY1004', 'Distributor', 217),
 ('WARE-UHY1004', 'Wholesale', 143),
 ('WARE-XYS1001', 'In-Store', 495),
 ('WARE-XYS1001', 'Online', 393),
 ('WARE-XYS1001', 'Distributor', 201),
 ('WARE-XYS1001', 'Wholesale', 133)]

#### What are the 5 most profitable products in terms of Unit Cost vs Unit Price?

In [77]:
%%sql

SELECT 
    "_ProductID", 
    AVG("Unit Price" - "Unit Cost") AS Average_Profit_Per_Unit
FROM 
    sales
GROUP BY 
    "_ProductID"
ORDER BY 
    Average_Profit_Per_Unit DESC
LIMIT 5;


 * postgresql://postgres:***@localhost:5432/salesdb
5 rows affected.


[(35, 967.8269642857142),
 (10, 967.1342321428577),
 (5, 944.5206878612717),
 (19, 942.9012134146336),
 (14, 934.4703742331294)]

#### 

#### Which products are more often sold online vs in-store?

In [78]:
%%sql

SELECT 
    "_ProductID", 
    "Sales Channel", 
    COUNT("OrderNumber") AS Total_Orders
FROM 
    sales
GROUP BY 
    "_ProductID", "Sales Channel"
ORDER BY 
    "_ProductID", Total_Orders DESC;


 * postgresql://postgres:***@localhost:5432/salesdb
188 rows affected.


[(1, 'In-Store', 74),
 (1, 'Online', 51),
 (1, 'Distributor', 24),
 (1, 'Wholesale', 13),
 (2, 'In-Store', 71),
 (2, 'Online', 51),
 (2, 'Distributor', 30),
 (2, 'Wholesale', 21),
 (3, 'In-Store', 72),
 (3, 'Online', 55),
 (3, 'Distributor', 26),
 (3, 'Wholesale', 19),
 (4, 'In-Store', 89),
 (4, 'Online', 58),
 (4, 'Distributor', 29),
 (4, 'Wholesale', 24),
 (5, 'In-Store', 63),
 (5, 'Online', 53),
 (5, 'Distributor', 34),
 (5, 'Wholesale', 23),
 (6, 'In-Store', 62),
 (6, 'Online', 57),
 (6, 'Distributor', 21),
 (6, 'Wholesale', 14),
 (7, 'In-Store', 64),
 (7, 'Online', 57),
 (7, 'Distributor', 25),
 (7, 'Wholesale', 20),
 (8, 'In-Store', 81),
 (8, 'Online', 53),
 (8, 'Distributor', 36),
 (8, 'Wholesale', 25),
 (9, 'In-Store', 73),
 (9, 'Online', 53),
 (9, 'Distributor', 31),
 (9, 'Wholesale', 13),
 (10, 'In-Store', 65),
 (10, 'Online', 54),
 (10, 'Wholesale', 25),
 (10, 'Distributor', 24),
 (11, 'In-Store', 79),
 (11, 'Online', 56),
 (11, 'Distributor', 25),
 (11, 'Wholesale', 22),
 (12, 'Online', 65),
 (12, 'In-Store', 63),
 (12, 'Distributor', 34),
 (12, 'Wholesale', 19),
 (13, 'In-Store', 66),
 (13, 'Online', 53),
 (13, 'Distributor', 36),
 (13, 'Wholesale', 20),
 (14, 'In-Store', 72),
 (14, 'Online', 47),
 (14, 'Distributor', 24),
 (14, 'Wholesale', 20),
 (15, 'In-Store', 61),
 (15, 'Online', 59),
 (15, 'Distributor', 20),
 (15, 'Wholesale', 17),
 (16, 'In-Store', 69),
 (16, 'Online', 45),
 (16, 'Distributor', 34),
 (16, 'Wholesale', 29),
 (17, 'In-Store', 85),
 (17, 'Online', 50),
 (17, 'Distributor', 27),
 (17, 'Wholesale', 16),
 (18, 'In-Store', 61),
 (18, 'Online', 52),
 (18, 'Distributor', 30),
 (18, 'Wholesale', 13),
 (19, 'In-Store', 69),
 (19, 'Online', 49),
 (19, 'Distributor', 23),
 (19, 'Wholesale', 23),
 (20, 'In-Store', 60),
 (20, 'Online', 51),
 (20, 'Distributor', 41),
 (20, 'Wholesale', 18),
 (21, 'In-Store', 70),
 (21, 'Online', 51),
 (21, 'Distributor', 31),
 (21, 'Wholesale', 16),
 (22, 'In-Store', 75),
 (22, 'Online', 47),
 (22, 'Distributor', 25),
 (22, 'Wholesale', 21),
 (23, 'In-Store', 87),
 (23, 'Online', 58),
 (23, 'Distributor', 35),
 (23, 'Wholesale', 14),
 (24, 'In-Store', 64),
 (24, 'Online', 51),
 (24, 'Distributor', 33),
 (24, 'Wholesale', 15),
 (25, 'In-Store', 79),
 (25, 'Online', 46),
 (25, 'Distributor', 37),
 (25, 'Wholesale', 18),
 (26, 'In-Store', 69),
 (26, 'Online', 37),
 (26, 'Distributor', 30),
 (26, 'Wholesale', 22),
 (27, 'In-Store', 88),
 (27, 'Online', 54),
 (27, 'Distributor', 26),
 (27, 'Wholesale', 22),
 (28, 'In-Store', 75),
 (28, 'Online', 57),
 (28, 'Distributor', 25),
 (28, 'Wholesale', 21),
 (29, 'In-Store', 71),
 (29, 'Online', 57),
 (29, 'Distributor', 33),
 (29, 'Wholesale', 15),
 (30, 'In-Store', 66),
 (30, 'Online', 51),
 (30, 'Distributor', 25),
 (30, 'Wholesale', 14),
 (31, 'In-Store', 65),
 (31, 'Online', 55),
 (31, 'Distributor', 31),
 (31, 'Wholesale', 18),
 (32, 'In-Store', 59),
 (32, 'Online', 40),
 (32, 'Distributor', 32),
 (32, 'Wholesale', 14),
 (33, 'In-Store', 73),
 (33, 'Online', 45),
 (33, 'Distributor', 31),
 (33, 'Wholesale', 17),
 (34, 'In-Store', 64),
 (34, 'Online', 55),
 (34, 'Distributor', 22),
 (34, 'Wholesale', 11),
 (35, 'In-Store', 69),
 (35, 'Online', 47),
 (35, 'Distributor', 31),
 (35, 'Wholesale', 21),
 (36, 'In-Store', 63),
 (36, 'Online', 48),
 (36, 'Distributor', 30),
 (36, 'Wholesale', 29),
 (37, 'In-Store', 96),
 (37, 'Online', 44),
 (37, 'Distributor', 35),
 (37, 'Wholesale', 25),
 (38, 'In-Store', 69),
 (38, 'Online', 60),
 (38, 'Distributor', 34),
 (38, 'Wholesale', 20),
 (39, 'Online', 60),
 (39, 'In-Store', 57),
 (39, 'Distributor', 25),
 (39, 'Wholesale', 22),
 (40, 'In-Store', 81),
 (40, 'Online', 58),
 (40, 'Distributor', 25),
 (40, 'Wholesale', 20),
 (41, 'In-Store', 61),
 (41, 'Online', 54),
 (41, 'Distributor', 43),
 (41, 'Wholesale', 21),
 (42, 'In-Store', 64),
 (42, 'Online', 40),
 (42, 'Distributor', 25),
 (42, 'Wholesale', 18),
 (43, 'In-Store', 74),
 (43, 'Online', 52),
 (43, 'Distributor', 22),
 (43, 'Wholesale',

### <center>RFM Analysis </center>

RFM (Recency, Frequency, Monetary) analysis is a customer segmentation technique that uses past purchase behavior to divide customers into groups.

1. Recency (R): Who have purchased recently? Number of days between latest order date and now for each customer.
2. Frequency (F): Who has purchased frequently? It means the total number of purchases.
3. Monetary Value(M): Who have high purchase amount? It means total money customer spent.

In [79]:
#

In [80]:
%%sql

SELECT 
    "_CustomerID",
    MAX("OrderDate") AS Latest_Order_Date,  -- Recency
    COUNT(DISTINCT "OrderNumber") AS Total_Orders,  -- Frequency
    SUM("Order Quantity" * "Unit Price") AS Total_Sales  -- Monetary
FROM 
    sales
GROUP BY 
    "_CustomerID";



 * postgresql://postgres:***@localhost:5432/salesdb
50 rows affected.


[(1, datetime.datetime(2020, 12, 23, 0, 0), 152, 1322278.5000000002),
 (2, datetime.datetime(2020, 12, 23, 0, 0), 135, 1346264.4999999998),
 (3, datetime.datetime(2020, 12, 22, 0, 0), 181, 1831947.5000000002),
 (4, datetime.datetime(2020, 12, 27, 0, 0), 167, 1770582.2000000007),
 (5, datetime.datetime(2020, 12, 2, 0, 0), 159, 1609232.8000000005),
 (6, datetime.datetime(2020, 12, 28, 0, 0), 143, 1530755.7000000002),
 (7, datetime.datetime(2020, 12, 29, 0, 0), 153, 1544061.8999999994),
 (8, datetime.datetime(2020, 12, 27, 0, 0), 142, 1411522.4999999998),
 (9, datetime.datetime(2020, 12, 24, 0, 0), 171, 1830038.0000000002),
 (10, datetime.datetime(2020, 12, 17, 0, 0), 158, 1763942.4999999995),
 (11, datetime.datetime(2020, 12, 26, 0, 0), 178, 1883410.1999999997),
 (12, datetime.datetime(2020, 12, 29, 0, 0), 210, 2248332.3999999994),
 (13, datetime.datetime(2020, 12, 28, 0, 0), 171, 1868321.8),
 (14, datetime.datetime(2020, 12, 27, 0, 0), 157, 1452225.0000000007),
 (15, datetime.datetime(2020, 12, 28, 0, 0), 142, 1512726.0000000007),
 (16, datetime.datetime(2020, 12, 29, 0, 0), 135, 1547773.7000000004),
 (17, datetime.datetime(2020, 12, 26, 0, 0), 175, 1962014.6000000006),
 (18, datetime.datetime(2020, 12, 26, 0, 0), 186, 1845783.0),
 (19, datetime.datetime(2020, 12, 29, 0, 0), 165, 1723534.8000000005),
 (20, datetime.datetime(2020, 12, 23, 0, 0), 167, 1523888.2000000007),
 (21, datetime.datetime(2020, 12, 29, 0, 0), 164, 1881627.9999999988),
 (22, datetime.datetime(2020, 12, 28, 0, 0), 140, 1494033.0),
 (23, datetime.datetime(2020, 12, 20, 0, 0), 164, 1793523.0000000007),
 (24, datetime.datetime(2020, 12, 9, 0, 0), 151, 1429960.8999999994),
 (25, datetime.datetime(2020, 12, 27, 0, 0), 162, 1762783.4000000001),
 (26, datetime.datetime(2020, 12, 21, 0, 0), 153, 1561783.4),
 (27, datetime.datetime(2020, 12, 29, 0, 0), 144, 1398008.5999999994),
 (28, datetime.datetime(2020, 12, 24, 0, 0), 145, 1408621.3999999997),
 (29, datetime.datetime(2020, 12, 30, 0, 0), 179, 2112221.9000000004),
 (30, datetime.datetime(2020, 12, 28, 0, 0), 159, 1640367.7),
 (31, datetime.datetime(2020, 12, 28, 0, 0), 152, 1450060.8999999994),
 (32, datetime.datetime(2020, 12, 30, 0, 0), 173, 1856355.5999999999),
 (33, datetime.datetime(2020, 12, 28, 0, 0), 156, 1895188.7999999996),
 (34, datetime.datetime(2020, 12, 17, 0, 0), 176, 1937827.6000000003),
 (35, datetime.datetime(2020, 12, 22, 0, 0), 145, 1555069.9999999988),
 (36, datetime.datetime(2020, 12, 28, 0, 0), 156, 1620904.2000000011),
 (37, datetime.datetime(2020, 12, 21, 0, 0), 152, 1622257.6),
 (38, datetime.datetime(2020, 12, 24, 0, 0), 150, 1413988.0999999999),
 (39, datetime.datetime(2020, 12, 28, 0, 0), 176, 1823941.0000000005),
 (40, datetime.datetime(2020, 12, 23, 0, 0), 150, 1636897.0999999999),
 (41, datetime.datetime(2020, 12, 30, 0, 0), 161, 1533047.1),
 (42, datetime.datetime(2020, 12, 30, 0, 0), 161, 1653814.6000000006),
 (43, datetime.datetime(2020, 12, 27, 0, 0), 151, 1564275.8),
 (44, datetime.datetime(2020, 12, 29, 0, 0), 156, 1529991.8999999994),
 (45, datetime.datetime(2020, 12, 27, 0, 0), 156, 1630029.5999999992),
 (46, datetime.datetime(2020, 12, 28, 0, 0), 157, 1508082.9000000001),
 (47, datetime.datetime(2020, 12, 24, 0, 0), 168, 1579625.500000001),
 (48, datetime.datetime(2020, 12, 22, 0, 0), 172, 1670108.9999999998),
 (49, datetime.datetime(2020, 12, 25, 0, 0), 152, 1503466.6000000006),
 (50, datetime.datetime(2020, 12, 16, 0, 0), 163, 1696225.5999999996)]

The insights obtained from this analysis can be used to segment customers based on their recency, frequency, and monetary value. This segmentation can help identify high-value customers, loyal customers, and customers who need to be targeted for re-engagement. It enables businesses to tailor their marketing strategies and offerings to specific customer segments, thereby maximizing customer satisfaction and profitability. Segmentation, is however, not covered in this analysis